snitun/server/run.py

```diff
@@ -106,9 +117,145 @@ async def _handler(
+                self._list_peer.handle_connection(reader, writer, data=data)
             )
+
+
+class SniTunServerWorker(Thread):
+    """SniTunServer helper class for Worker."""
+
+    def __init__(
+        self,
+        fernet_keys: List[str],
+        host: Optional[str] = None,
+        port: Optional[int] = None,
+        worker_size: Optional[int] = None,
+        throttling: Optional[int] = None,
+    ):
+        """Initialize SniTun Server."""
+        super().__init__()
+
+        self._host: Optional[str] = host
+        self._port: int = port or 443
+        self._fernet_keys: List[str] = fernet_keys
+        self._throttling: Optional[int] = throttling
+        self._worker_size: int = worker_size or (cpu_count() * 2)
+        self._workers: List[ServerWorker] = []
+        self._running: bool = False
+
+        # TCP server
+        self._server: Optional[socket.socket] = None
+        self._poller: Optional[select.epoll] = None
+
+    def start(self) -> None:
+        """Run server."""
+        # Init first all worker, we don't want the epoll on the childs
+        for _ in range(self._worker_size):
+            worker = ServerWorker(self._fernet_keys, throttling=self._throttling)
+            worker.start()
+            self._workers.append(worker)
+
+        self._server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._server.bind((self._host, self._port))
+        self._server.setblocking(False)
+        self._server.listen(80 * 1000)
+
+        self._running = True
+        self._poller = select.epoll()
+        self._poller.register(self._server.fileno(), select.EPOLLIN)
+
+        super().start()
+
+    def stop(self) -> None:
+        """Stop server."""
+        self._running = False
+        self.join()
+
+        # Shutdown all workers
+        for worker in self._workers:
+            worker.shutdown()
+            worker.close()
+
+        self._workers.clear()
+        self._server.close()
+        self._poller.close()
+
+    def run(self) -> None:
+        """Handle incoming connection."""
+        fd_server = self._server.fileno()
+        connections: Dict[int, socket.socket] = {}
+        worker_lb = cycle(self._workers)
+        stale: Dict[int, int] = {}
+
+        while self._running:
+            events = self._poller.poll(1)
+            for fileno, event in events:
+                # New Connection
+                if fileno == fd_server:
+                    con, _ = self._server.accept()
+                    con.setblocking(False)
+
+                    self._poller.register(
+                        con.fileno(), select.EPOLLIN | select.EPOLLHUP | select.EPOLLERR
+                    )
+                    connections[con.fileno()] = con
+                    stale[con.fileno()] = 0
+
+                # Read hello & forward to worker
+                elif event & select.EPOLLIN:
+                    self._poller.unregister(fileno)
+                    con = connections.pop(fileno)
```

Comment:
You're using both `con.fileno()` and `fileno` interchangeably. Let's stick to only using `fileno`.

```suggestion
                    self._poller.register(
                        fileno, select.EPOLLIN | select.EPOLLHUP | select.EPOLLERR
                    )
                    connections[fileno] = con
                    stale[fileno] = 0

                # Read hello & forward to worker
                elif event & select.EPOLLIN:
                    self._poller.unregister(fileno)
                    con = connections.pop(fileno)
```

Reply:
That will not work. The fileno is the server and we need to create a new file handler (socket) for the new connection and push that into our connection manager. This is a special case if new connection come in and we need call accept() to get the new socket object. If you run the tests, you will see that is not working

Reply:
we use all over the place con for socket/connection and fileno for his file descriptor reference inside the kernel

Reply:
Ok, so if you add things to the `connections` dictionary, you use `con.fileno()` and while popping it, you use `fileno`. Doesn't that mean they are the same?

Reply:
correct

Reply:
So that was what I was trying to say Let's just use `fileno` instead of `con.fileno()` to make it easier to understand the code.

Reply:
And I try to say that fileno is not each time the same as con.fileno(). I agree that we should use the exiting fileno if it's possible